### PR TITLE
Fix HostToGpuCoalesceIterator sometimes closing input batches

### LINKED ISSUE #6995: [BUG] HostToGpuCoalesceIterator can sometimes close input batches
The host batches have the opposite ownership rules of the GPU batches, where the producer needs to close the batch rather than the consumer.  HostToGpuCoalesceIterator can sometimes violate this by closing an input batch (e.g.: on an empty batch) because it is leveraging AbstractGpuCoalesceIterator which makes some assumptions on batch ownership that do not apply when the columnar batches are on the host.

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #6995.  Added a `cleanupInputBatch` callback method that allows host-based coalesce iterators to avoid closing the input batches to match the expected ownership semantics.

### CODE REVIEW

**Comment by jlowe** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/HostCoalesceBatchesSuite.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+package com.nvidia.spark.rapids
+
+import org.scalatest.FunSuite
+
+class HostCoalesceBatchesSuite extends FunSuite with Arm {
+  test("close handling of host batches") {
+
+    // Host columnar batches are closed by the producer, not the consumer.
+    val iterchain = AutoCloseColumnBatchIterator
```
Comment (abellina) [inline tests/src/test/scala/com/nvidia/spark/rapids/HostCoalesceBatchesSuite.scala:25]: was this committed by mistake (perhaps missing a diff)?

File: tests/src/test/scala/com/nvidia/spark/rapids/GpuCoalesceBatchesSuite.scala [hunk 2]
```diff
@@ -233,6 +233,39 @@ class GpuCoalesceBatchesSuite extends SparkQueryCompareTestSuite {
...
+        streamTime = NoopMetric,
+        concatTime = NoopMetric,
+        copyBufTime = NoopMetric,
+        semTime = NoopMetric,
+        opTime = NoopMetric,
+        peakDevMemory = NoopMetric,
+        opName = "concat test",
+        useArrowCopyOpt = false)
+    coalesceIter.foreach { batch =>
+      withResource(batch) { _ =>
```
Comment (abellina) [inline tests/src/test/scala/com/nvidia/spark/rapids/GpuCoalesceBatchesSuite.scala:262]: nit, a comment above `withResource` may be nice saying we are making sure this doesn't double-close.

**Comment by jlowe** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/HostCoalesceBatchesSuite.scala (hunk 1 above)
Comment (jlowe) [reply tests/src/test/scala/com/nvidia/spark/rapids/HostCoalesceBatchesSuite.scala]: Oops, cleaned this up.

File: tests/src/test/scala/com/nvidia/spark/rapids/GpuCoalesceBatchesSuite.scala (hunk 2 above)
Comment (jlowe) [reply tests/src/test/scala/com/nvidia/spark/rapids/GpuCoalesceBatchesSuite.scala:263]: Added a comment.

**Comment by jlowe** [comment]
build

**Review by abellina — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

**Pull request merged**